### ISSUE TITLE
cert issue / refused to connect

### ISSUE THREAD

**Issue by drewfitz10** [laravel/valet#654]
I am getting the `refused to connect` screen. I can ping the sites no problem and get response.  I believe it may have to do with the cert issued by `valet secure`. 

When I run sudo `/usr/local/opt/nginx/bin/nginx -g 'daemon off;'` I receive the following response: 

`nginx: [emerg] PEM_read_bio_X509_AUX("/Users/<user>/.config/valet/Certificates/site.app.crt") failed (SSL: error:0906D06C:PEM routines:PEM_read_bio:no start line:Expecting: TRUSTED CERTIFICATE)`

Recently due to company requirements I had to change my user folder name and path in Mac OS Mojave. Prior to change had no issues with valet. I have followed every error issue to try and resolve with no luck.

**Comment by drbyte** [comment]
Things I'd try:

- `valet tld test` (or whatever TLD you want) This will rebuild all the certs.
- `valet unsecure` and `valet secure`, to rebuild that specific cert
- `cat /Users/<user>/.config/valet/Certificates/site.app.crt` to see what's in the cert. What's the starting line actually say, vs the expected text indicated by the error message?

**Comment by drewfitz10** [comment]
Followed the steps provided and the `.app.crt` file is empty. `cat` does not print anything to the console

**Comment by drbyte** [comment]
If you run `valet unsecure`, can you access the sites via http (not https)? 
If not, then maybe the templates that valet builds things from are damaged?

Did you run `valet install` in the new user account?

Valet is not well-suited for running from multiple different user accounts on the same Mac.
Maybe it'd be worth doing a full uninstall of valet and redo things in the new user acct?

**Comment by drewfitz10** [comment]
I can get sites to serve from the `.test` domain, but only unsecured. I have done full uninstall and reinstalled multiple times now without any luck.

**Comment by drbyte** [comment]
The `.test` domains would use a `domain.test.crt` file, not `domain.app.crt`.

Perhaps you've got your setup confused by mixing TLDs? Valet only supports a single TLD at a time.

**Comment by drewfitz10** [comment]
Apologies for the confusion. Even when securing `.test` the `.crt` file is empty. I tested both.

`.test` works unsecured. Does not work secured
`.app` does not work unsecured or secured.

**Comment by drbyte** [comment]
You said you had to change the user path. I"m not sure exactly what steps you took nor the why behind them. Renaming/relocating folders can change a lot of internal pointers that mess with things that depend on them. 

Valet requires that your valet-related configuration is stored in `~/.config/valet`. It expects (but doesn't enforce) that the `~` part of that resolves to `/Users/<user>/`. (Actually, while Valet doesn't enforce it, the nginx stubs valet uses probably do.)

It's possible that you've changed things in a way that valet no longer can write (or read) files from (all of) the expected locations.

**Comment by drbyte** [comment]
If it's only related to the certificates, you could go down this path:
There's a lot going on in the code that builds the certificates.
Here's the section of code: https://github.com/laravel/valet/blob/master/cli/Valet/Site.php#L223-L418
You might want to look into some of the commands it runs to see whether there are dependencies that have been broken during your change.

**Comment by drewfitz10** [comment]
After some digging through the code you referenced I was able to hack a solution, but still not sure why it is failing.

In this section of code: https://github.com/laravel/valet/blob/fcaa484ee3a9d852456e5bd1775648a603d9ebed/cli/Valet/Site.php#L307-L312

If I run both of those command manually, the creation of the `.crt` and adding the `.crt` to keychain, then `valet restart` I can server to any tld over https.

Any thoughts on why that creation of the `.crt` would be getting corrupted? It gets created, but can't get imported into keychain and has an empty file contents. Running manually presents no problem.

**Comment by drewfitz10** [comment]
If I run the scripts myself the import to keychain is fine, so I believe it is permission issues with the creation of the certificate. The certificate gets created, but it is empty

**Comment by drbyte** [comment]
Agreed: it seems like some sort of permission issue resulting from whatever the changes were that you made to your user account.

**Comment by drbyte** [comment]
@michaelroper Please explain your environment. Start with assuming I have just unboxed a new Mac and can install any OSX version (and perform upgrades from older versions if thats how you got to your current version). How do I set it up so I can recreate your symptoms?

**Comment by mpatzner** [comment]
I've got the same problem as mentioned by @michaelroper and @drewfitz10. Sites work via http, but not via https.

My tld is "test" and I have recently run "brew upgrade" and reinstalled Valet by removing it's global composer package and running valet uninstall. I'm on Valet 2.1.1, Composer 1.7.2 and Homebrew 1.7.7 using macOS Mojave (upgraded last night, but the issue appeared on the previous macOS version as well).

If it helps, here's the output of "brew services list":

```
dnsmasq   started root               /Library/LaunchDaemons/homebrew.mxcl.dnsmasq.plist
httpd     started root               /Library/LaunchDaemons/homebrew.mxcl.httpd.plist
mysql@5.7 started mpatzner /Users/mpatzner/Library/LaunchAgents/homebrew.mxcl.mysql@5.7.plist
nginx     started root               /Library/LaunchDaemons/homebrew.mxcl.nginx.plist
php       started root               /Library/LaunchDaemons/homebrew.mxcl.php.plist
redis     started root               /Library/LaunchDaemons/homebrew.mxcl.redis.plist
```

**Comment by ylving** [comment]
Hey

I recently began experiencing the same issue. What causes it I do not know yet, but the symptoms are the same; "Valet secure" generates an empty CRT file, nginx does not start listening on port:443 and it even stops listening on port:80. Which explains why none of the local sites are responding.

I did however find a workaround until the bug is fixed if that is of any help.

1. `valet secure` generates a conf-file in addition to the crt-, csr- and key-file in the Certificates folder.
2. Run openssl on this conf-file: `sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout sitename.test.key -out sitename.test.crt -config sitename.test.conf`.
3. Restart nginx `sudo brew services restart nginx`

Nginx now serves that site over https and all the others over http as expected.

**Comment by ylving** [comment]
Hi

Same issue with an empty *.crt being created from `valet secure` and at the same time nginx stops listening on port:80 (and never starts listening on port:443).

In my case it needed to run the `openssl` command as sudo.

Changing `$this->cli->runAsUser` to `$this->cli->run` on line 307 in Site.php fixed the issue for me.
https://github.com/laravel/valet/blob/fcaa484ee3a9d852456e5bd1775648a603d9ebed/cli/Valet/Site.php#L307-L310

**Comment by drbyte** [comment]
@ylving wrote:
> Changing `$this->cli->runAsUser` to `$this->cli->run` on `line 307` in `Site.php` fixed the issue.

I just tested this on older versions: High Sierra (fresh-ish install) and El Capitan (old install) without any problems. Granted, some stuff was already in the KeyChain. But re-generation of certificates worked without issue.

**Comment by drbyte** [comment]
While I still can't trigger the empty-crt file on this Mojave install, 
I'd like to explore making the small change that @ylving suggested: 
https://github.com/laravel/valet/issues/654#issuecomment-439302720
Simply edit `~/.composer/vendor/laravel/valet/cli/Valet/Site.php` line 307 - change `runAsUser` to just `run`
and then try securing your site/s again.

### Previous posts:
Since @drewfitz10 @michaelroper @mpatzner @ViktorPontinen posted as having the problem, it'd be great to get feedback whether this helps in your case!
I'd ALSO like to know whether valet is prompting you for a password when you run `valet secure`. I wonder if running `valet trust` not only stops prompting but also solves your .crt issue without the code change on line 307.

**Comment by ylving** [comment]
`valet secure` do prompt me for my password both before and after running `valet trust`.
Unfortunately it does not fix the .crt issue in my case.

I am running the same versions as ViktorPontinen:
osx: 10.14.1
php: 7.2.12
nginx: 1.15.6
dnsmasq: 2.80
valet: 2.1.3

**Comment by drbyte** [comment]
@ylving wrote:
> valet secure do prompt me for my password both before and after running valet trust.

Does making the change I suggested below help fix that?
https://github.com/laravel/valet/issues/648#issuecomment-427672783

**Comment by ylving** [comment]
Unfortunately no. I went through the thread, checked the files and restarted for good measure, but no change. I´ll dump the contents below if that is of any help.

My sudoer already included the missing part referenced in #648 
```
#
# Sample /etc/sudoers file.
#
# This file MUST be edited with the 'visudo' command as root.
#
# See the sudoers man page for the details on how to write a sudoers file.

##
# Override built-in defaults
##
Defaults	env_reset
Defaults	env_keep += "BLOCKSIZE"
Defaults	env_keep += "COLORFGBG COLORTERM"
Defaults	env_keep += "__CF_USER_TEXT_ENCODING"
Defaults	env_keep += "CHARSET LANG LANGUAGE LC_ALL LC_COLLATE LC_CTYPE"
Defaults	env_keep += "LC_MESSAGES LC_MONETARY LC_NUMERIC LC_TIME"
Defaults	env_keep += "LINES COLUMNS"
Defaults	env_keep += "LSCOLORS"
Defaults	env_keep += "SSH_AUTH_SOCK"
Defaults	env_keep += "TZ"
Defaults	env_keep += "DISPLAY XAUTHORIZATION XAUTHORITY"
Defaults	env_keep += "EDITOR VISUAL"
Defaults	env_keep += "HOME MAIL"

Defaults	lecture_file = "/etc/sudo_lecture"

##
# User alias specification
##
# User_Alias	FULLTIMERS = millert, mikef, dowdy

##
# Runas alias specification
##
# Runas_Alias	OP = root, operator

##
# Host alias specification
##
# Host_Alias	CUNETS = 128.138.0.0/255.255.0.0
# Host_Alias	CSNETS = 128.138.243.0, 128.138.204.0/24, 128.138.242.0
# Host_Alias	SERVERS = master, mail, www, ns
# Host_Alias	CDROM = orion, perseus, hercules

##
# Cmnd alias specification
##
# Cmnd_Alias	PAGERS = /usr/bin/more, /usr/bin/pg, /usr/bin/less

##
# User specification
##

# root and users in group wheel can run anything on any machine as any user
root		ALL = (ALL) ALL
%admin		ALL = (ALL) ALL
my.username	ALL = (ALL) ALL
## Read drop-in files from /private/etc/sudoers.d
## (the '#' here does not indicate a comment)
#includedir /private/etc/sudoers.d


```

In `/etc/sudoers` there are 2 files: `brew` and `valet`
The content of `valet` is:
`Cmnd_Alias VALET = /usr/local/bin/valet *
%admin ALL=(root) NOPASSWD: VALET`

Running `id` outputs:
`uid=1469598454(my.username) gid=2011178512 groups=2011178512,12(everyone),20(staff),62(netaccounts),701(com.apple.sharepoint.group.2),703(com.apple.sharepoint.group.3),702(com.apple.sharepoint.group.1)`

**Comment by ViktorPontinen** [comment]
> While I still can't trigger the empty-crt file on this Mojave install,
> I'd like to explore making the small change that @ylving suggested:
> [#654 (comment)](https://github.com/laravel/valet/issues/654#issuecomment-439302720)
> Simply edit `~/.composer/vendor/laravel/valet/cli/Valet/Site.php` line 307 - change `runAsUser` to just `run`
> and then try securing your site/s again.
> 
> ### Previous posts:
> Since @drewfitz10 @michaelroper @mpatzner @ViktorPontinen posted as having the problem, it'd be great to get feedback whether this helps in your case!
> I'd ALSO like to know whether valet is prompting you for a password when you run `valet secure`. I wonder if running `valet trust` not only stops prompting but also solves your .crt issue without the code change on line 307.

Valet does not ask for password when I run `valet secure`.
Running `valet trust` before `valet secure` did not solve the empty .crt.

However, changing the code on line 307 resolved the issue for me and allowed a non-empty .crt to be generated.

**Comment by drewfitz10** [comment]
@drbyte I can confirm changing line 307 to just `run` worked and resolved the issue. I had previously run `Valet trust` which stopped the prompt for password, but didn't solve the issue.

**Comment by mattstauffer** [comment]
I ran into this issue with a friend's computer. It was definitely something permissions related. My only smell was that, when he installed something unrelated earlier, we had to use sudo, and it gave me the sense that he may have used sudo to install something else via Homebrew earlier, and there must've been some access control issues somewhere in the directory where the crt was created.

If I had more time, I would dig into the ownership and permissions of the .crt's parent folders on his machine vs mine to see if they were any different.

**Comment by eithed** [comment]
On my dev machine crt file is empty whenever I install valet / try to secure my dev directory. The error that is reported by nginx is: `PEM_read_bio_X509_AUX("~/.config/valet/Certificates/Certificates.test.crt") failed (SSL: error:0906D06C:PEM routines:PEM_read_bio:no start line:Expecting: TRUSTED CERTIFICATE)`

To fix this I run the steps outlined here to generate self-signed certificate: https://www.digitalocean.com/community/tutorials/openssl-essentials-working-with-ssl-certificates-private-keys-and-csrs

```
openssl req \
       -newkey rsa:2048 -nodes -keyout Certificates.test.key \
       -x509 -days 365 -out Certificates.test.crt
```

**Comment by eithed** [comment]
@drbyte 

1. My dev directory is ~/Sites - as far as I remember I never configured anything to look for Certificates folder and just assumed that's where valet stores certificates for domains; if I remember correctly as soon as I run `valet secure` for the first time that's the error I started to get in nginx logs and I couldn't make it go away, even when running `valet unsecure`; only setting up certificates made it go away (however the content is still served via http).
2. I don't know how to answer that - I've setup my local env per installation manual, then migrated from one machine to another (which indeed messed up the ownership of the files, but I uninstalled / unlinked / installed everything again). I've parked my dev folder and setup TLD to be test - I think that's the only steps I've done outside installation steps.

Today I've had to reinstall my valet as I had to revert to PHP 7.2, which made me look at the issue again and found this thread.

**Comment by drbyte** [comment]
@eithed thanks, the added context helps.

Yes, the "Certificates folder is where valet stores certificates, but it never creates in there a file called "Certificates.test.crt" unless someone runs `valet secure Certificates` (or equiv inside a dev directory named "Certificates")

This is the first I'm seeing this captured information from nginx logs, so thanks for that.

**Comment by RomainGoncalves** [comment]
> Changing `$this->cli->runAsUser` to `$this->cli->run` on line 307 in Site.php fixed the issue.

This helped me today as well. Had the same issue of my site having a connection refused when doing `valet secure`, while the unsecured version worked.

Changing `Site.php` at: `~/.composer/vendor/laravel/valet/cli/Valet/Site.php` then running `valet secure` again did the trick for me.

**Comment by bilfeldt** [comment]
I have experienced the same problem. It started after upgrading to macOS 10.14.2 and persistent after upgrading to 10.14.2. I tried completely removing all brew services but in the end, it was fixed by:

> Changing `$this->cli->runAsUser` to `$this->cli->run` on line 307 in Site.php

as [stated](https://github.com/laravel/valet/issues/654#issuecomment-439302720) earlier.

**Comment by drbyte** [comment]
Reference Note: The fix mentioned above was released in v2.2.2

**Comment by drbyte** [comment]
@mattstauffer I believe this can be closed.
